pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -937,38 +1055,66 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 		p.sendRefused(scopedLog, w, request)
 		return
 	}
+
 	stat.ProcessingTime.End(true)
 	stat.UpstreamTime.Start()
 
-	dialer := net.Dialer{
-		Timeout:   2 * time.Second,
-		LocalAddr: originalSrcAddr,
-		Control: func(network, address string, c syscall.RawConn) error {
-			var soerr error
-
-			if err := c.Control(func(su uintptr) {
-				soerr = setSoMarks(int(su), ipFamily, ep.GetIdentity())
-			}); err != nil {
-				return err
-			}
-			return soerr
-		},
+	// locate client to re-use if possible.
+	request.Id = dns.Id() // force a random new ID for this request
+	key := protocol + "-" + originalSrcAddr.String() + "-" + targetServerAddrStr
+	p.Lock()
+	client := p.DNSClients[key]
+	if client == nil {
```

Comment:
It would be nice to extract this whole block as `NewDNSClient` function.

Reply:
+1 from me

Reply:
Done, also for `GetDNSClient` function, turned out pretty nice :-)